clang/lib/Frontend/CompilerInvocation.cpp

```diff
@@ -1342,23 +1342,51 @@ createBaseFS(const FileSystemOptions &FSOpts, const CASOptions &CASOpts,
   if (!CAS)
     return makeEmptyCASFS();
 
-  StringRef RootIDString = FSOpts.CASFileSystemRootID;
+  auto IsIncludeTreeFS = !FEOpts.CASIncludeTreeID.empty();
+
+  StringRef RootIDString =
+      IsIncludeTreeFS ? FEOpts.CASIncludeTreeID : FSOpts.CASFileSystemRootID;
+
   Expected<llvm::cas::CASID> RootID = CAS->parseID(RootIDString);
   if (!RootID) {
     llvm::consumeError(RootID.takeError());
     Diags.Report(diag::err_cas_cannot_parse_root_id) << RootIDString;
     return makeEmptyCASFS();
   }
 
-  Expected<std::unique_ptr<llvm::vfs::FileSystem>> ExpectedFS =
-      llvm::cas::createCASFileSystem(std::move(CAS), *RootID);
+  auto makeIncludeTreeFS = [&](std::shared_ptr<llvm::cas::ObjectStore> CAS,
+                              llvm::cas::CASID &ID)
+      -> Expected<IntrusiveRefCntPtr<llvm::vfs::FileSystem>> {
+    Optional<llvm::cas::ObjectRef> Ref = CAS->getReference(ID);
+    if (!Ref) {
+      Diags.Report(diag::err_cas_filesystem_cannot_be_initialized)
+          << RootIDString;
+      return makeEmptyCASFS();
```

Comment:
This looks like it duplicates the code below on line 1383 (`if (!ExpectedFS) {`)